### PR TITLE
Base value modernisation part 4

### PR DESCRIPTION
**Pull request by cdesouza-chromium**
Resolves https://github.com/brave/brave-browser/issues/24315

These are additional base::Value modernisation changes, to eliminate the use of deprecated methods/interfaces from brave code.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by tmancey — APPROVED** [review]
ads LGTM

File: vendor/bat-native-ledger/src/bat/ledger/internal/attestation/attestation_desktop.cc [hunk 1]
```diff
@@ -32,24 +32,20 @@ type::Result AttestationDesktop::ParseClaimSolution(
     return type::Result::LEDGER_ERROR;
   }
 
-  base::DictionaryValue* dictionary = nullptr;
-  if (!value->GetAsDictionary(&dictionary)) {
-    return type::Result::LEDGER_ERROR;
-  }
-
-  const auto* id = dictionary->FindStringKey("captchaId");
+  base::Value::Dict& dict = value->GetDict();
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/attestation/attestation_desktop.cc:35]: Can we use `const auto& dict` here and elsewhere?

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_util.cc [hunk 2]
```diff
@@ -207,45 +207,33 @@ std::string ConvertRewardTypeToString(const type::RewardsType type) {
...
-    const std::string& body,
-    base::Value* credentials) {
-  DCHECK(credentials);
-
+    const std::string& body) {
+  base::Value::List credentials;
   for (auto& item : token_list) {
-    base::Value token(base::Value::Type::DICTIONARY);
-    bool success;
+    absl::optional<base::Value::Dict> token;
```
Comment (szilardszaloki) [inline vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_util.cc:215]: How about:
```
if (auto token =
        ledger::is_testing
            ? GenerateSuggestionMock(item.token_value, item.public_key, body)
            : GenerateSuggestion(item.token_value, item.public_key, body)) {
  credentials.Append(std::move(*token));
}
```

**Review by zenparsing — APPROVED** [review]

File: vendor/bat-native-ledger/src/bat/ledger/internal/attestation/attestation_desktop.cc (hunk 1 above)
Comment (cdesouza-chromium) [reply vendor/bat-native-ledger/src/bat/ledger/internal/attestation/attestation_desktop.cc]: Done for all occurrences in all commits. Apologies I won't be able to produce them all in a fixup commit as I had to edit multiple commits to make this suggestion consistent across.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/bitflyer/get_balance/get_balance_bitflyer.cc [hunk 3]
```diff
@@ -55,25 +55,22 @@ type::Result GetBalance::ParseBody(const std::string& body, double* available) {
...
+  const auto* inventory = dict.FindList("inventory");
   if (!inventory) {
     BLOG(0, "Missing inventory");
     return type::Result::LEDGER_ERROR;
   }
 
-  for (const auto& item : inventory->GetList()) {
-    const auto* currency_code = item.FindStringKey("currency_code");
+  for (const auto& item : *inventory) {
+    DCHECK(item.is_dict());
```
[the remark this answers is not in the document]
Comment (cdesouza-chromium) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/bitflyer/get_balance/get_balance_bitflyer.cc]: Fixed in 9a9861ca6b708bf7f591b32cc7e9aa3864f237a5.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/promotion/get_available/get_available.cc [hunk 4]
```diff
@@ -85,50 +85,48 @@ type::Result GetAvailable::ParseBody(
...
   }
 
-  const auto promotion_size = promotions->GetList().size();
-  for (auto& item : promotions->GetList()) {
+  const auto promotion_size = promotions->size();
+  for (auto& value : *promotions) {
     type::PromotionPtr promotion = type::Promotion::New();
 
-    const auto* id = item.FindStringKey("id");
+    DCHECK(value.is_dict());
```
[the remark this answers is not in the document]
Comment (cdesouza-chromium) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/promotion/get_available/get_available.cc]: Fixed in 22f1ce4671e9e66d5d06f7a0ebab149356d95054.

**Review by emerick — APPROVED** [review]

**Pull request merged**